### DISCUSSION TITLE
Would you be disappointed if TShock didn't work on arm64/Raspberry Pi?

### DISCUSSION THREAD

**Discussion by hakusaro** [Pryaxis/TShock#2562] [General]
Hi! Currently, @SignatureBeef has a branch of TShock that works on .NET6. However, it depends on MonoMod hooks that may not work correctly, if at-all, on arm64/arm64e/arm in general. What this means is that TShock would essentially drop support for these environments going forward.

Would this disappoint you?

For example:
* TShock wouldn't work natively on Apple Silicon. You'd have to use Rosetta 2 to run TShock via .NET Core.
* TShock wouldn't work correctly on Raspberry Pi (all versions). You'd have to use x86_64 processors instead, like an Intel NUC.
* TShock wouldn't work correctly on ec2 using AWS Graviton or Graviton 2 instances.
* TShock wouldn't work correctly on Windows-on-arm (very limited availability at the moment).

If this would upset you, please make your voice as **loud and prominent as possible** so that we don't accidentally pull the plug on something you depend on.

See also: [Twitter poll](https://twitter.com/Pryaxis/status/1493100154362028032)

Edit: [An announcement](https://github.com/Pryaxis/TShock/discussions/2571) has been made on the future of TShock + Raspberry Pi / arm.

**Comment by ghost** [comment]
I would be very disappointed if support for arm64 is dropped, specifically for Raspberry Pi.

Raspberry Pi is almost the perfect platform for running a Terraria server, for those on a budget - in terms of both initial cost and energy usage. The cheapest x86_64 machines, such as Intel NUCs or machines with old CPUs, cost several times more than the $35 Raspberry Pi 4B (2GB RAM) and $55 Raspberry Pi 4B (4GB ram) models. The only x86_64 platforms that can rival the Pis' power consumption are single-board computers that also come at several times the initial cost of a Raspberry Pi (Model 4B is rated for 15 W _maximum_, achievable at max load and with some USB devices connected).

Personally, I have a TShock server running on a Raspberry Pi 4B 4GB, alongside several lightly-loaded web servers on a home lab. I just need it to handle four concurrent players, and the Pi is more than capable.

**Reply by hakusaro** [reply]
Are you using the RPi _specifically_ for Terraria, or in-addition to other tasks? Are you using the bare metal, or running through Docker?

Thank you for your power consumption notes, particularly. That's a very good point.

A cursory glance at x86_64 SBCs is...not good, as one might assume ($109ish from one manufacturer).

**Reply by carsmiley** [reply]
Hi,

Thanks a bunch for making and maintaining Tshock in the first place.

Like testplayername, I also use the Pi for running a Terraria server and I would be dissppointed to lose access to Tshock. I use the Pi 4B 8GB, and agree with everything he says.

The Pi is more than capable of running the half a dozen Docker-containers (nginx, vaultwarden, Home-Assistant, Zabbix, PiHole, Nextcloud) and still run Tshock (not in Docker) with 4 players while being enormously energy-efficient. Albeit that most of the containers don’t see continuous load. 

I also have looked at getting an x64-device specifically for projects not supporting arm. But it never seemed worth it for me since they are few, and usually require much beefier hardware than what a Pi or equivalent could offer.

**Reply by ghost** [reply]
I am not using Docker, but pretty close. My Pi is running a LXD server with a few containers, one of which is dedicated for running TShock. The Pi itself is not dedicated for TShock; the lightly-loaded web servers I mentioned are running on the same physical Pi. When someone is on the server, TShock takes the most sustained CPU cycles and RAM, with my reverse proxy (running in a different container but same Pi) taking a very distant second place. With a max of four players, dedicating a 4GB Pi for TShock seems like overkill.

carsmiley is running a much heavier load than me, although he has twice the RAM as well.

**Reply by pcider** [reply]
Hi,

I use the Raspberry Pi 4 2GB for TShock, It fits well for playing on the large world with my small group of 4 friends. Being an amateur at Linux, I have learnt a lot about working with bash scripts, GNU utilities and server management in general on linux with TShock on the Raspberry Pi. Without TShock on the Raspberry Pi, I wouldn't have the incentive to learn about these powerful tools just to setup a Terraria server for my friends playing on mobile. Being a novice computer engineering student, I feel that you may be potentially cutting out a part of the community who love using the Raspberry Pi in their projects, even if it's just cutting out few people or even only me.

**Reply by adriantomic** [reply]
I'm running Tshock on a dedicated raspberry pi 4. It works great, and it would be unfortunate if support for the RPi ended. Thanks for your hard work on TShock.

**Comment by arth3mis** [comment]
I have incredible fun with my friends on a Raspberry Pi server, the Pi is online 24/7 just for Terraria. The flexibility for a small party to just hop on from anywhere and play, without me having to worry about skyrocketing electronic bills, is just...so cool.

I have a Pi 3B+, so just arm32, and it's still working really well. It even handles 6 players (with only minor lag). I'm in awe of TShock's functionality and adjustability, and together with a Pi they are an iconic duo in my mind.

**Comment by matteobin** [comment]
I am a Pi owner as well.
Since Terraria 1.4 released I have been waiting for a new stable version of TShock to play with my friends.
Please, do not follow this path: Terraria and Pi, as someone has already pointed out, is a match made in heaven, because one is a little nice game that is actually complex and the other is a cheap small computer that in the right hands can be a good entry level server. These two things provide a similar joy for discovery and learning both in gaming and IT department. Terraria is a creative game and Pis are computers for creative people or better said for people who enjoy tinkering, like modders who enjoy creating new game experiences.

**Comment by Wuzado** [comment]
Another platform that's really nice to host on and would be gone after the transition is Oracle's Ampere A1 VMs, with a generous free tier of 4 OCPUs and 24 GBs of RAM. A lot of people, including myself, are running a server on this platform which has more than enough resources to run a Terraria server or even several of them.

**Reply by hakusaro** [reply]
Oh wow. That’s new! I haven’t heard of this. Thanks!

**Reply by FlyGuyGo** [reply]
I use this to host my tshock servers as well. It is amazing because it allows me to focus on making the game fun instead of collecting donations. I'd also venture to say that while consumers may have x86 for a while the server space is going in the direction of ARM. So adoption of ARM for tshock will only grow. Even without the Oracle free tier, ARM instances are cheaper to obtain because they consume less power. That is why Amazon's ARM instances are significantly cheaper than their x86 counterpart.

**Reply by hakusaro** [reply]
What's the cheapest graviton instance that runs TShock?

**Reply by Wuzado** [reply]
So AWS _incredibly_ confuses me, and thus I am not sure if the info I'm saying is correct/optimal, but it seems like the smallest Graviton instance listed, `m6g.medium`, has 1 vCPU and 4 GB of RAM, which seems like enough for hosting TShock. According to AWS's calculator, such VM would cost around 20 USD a month with a 1 year reservation plan and no upfront.

**Reply by FlyGuyGo** [reply]
I brought up AWS assuming that one day Oracle might not offer their free plan anymore and to just speak generally about the shift to ARM in the server space. However, as of now their free plan is definitely the best way to go. I know one might be skeptical but there really is no strings attached in my experience of using it for several months.

**Comment by funkiii** [comment]
Would be a shame if my Raspberry Pi Terraria Server would stop working.

**Comment by CtrlJames** [comment]
I would be upset, as I'm currently running several Terraria /TShock servers on Raspberry Pi's. Please (Pretty Please with strawberry on top) don't remove the ARM support!!!

**Comment by FlyGuyGo** [comment]
Repost of my comment just so it doesn't get missed and to be "loud and prominent as possible". I use Oracle's free Tier to host my tshock servers like @Wuzado and many others do. It is amazing because it allows me to focus on making the game fun instead of collecting donations. I'd also venture to say that while consumers may have x86 for a while the server space is going in the direction of ARM. So adoption of ARM for tshock will only grow. Even without the Oracle free tier, ARM instances are cheaper to obtain because they consume less power. That is why Amazon's ARM instances are significantly cheaper than their x86 counterpart.

**Comment by tylerjwatson** [comment]
Whoa before we do anything brash like drop architecture support, whats the technical limitations of these hooks, and can they be ported to arm? Is this just a case of effort? I’d like to have a look.

**Reply by SignatureBeef** [reply]
In short yes, it's possible to implement them given time and skill, or we can work around it by dropping the monomod runtime hook dependency. 
The main issue we face on arm64 is instance method detours simply don't invoke, however static detours do.
I haven't had much time to continue testing and debugging on my RPI using monomod itself or finding a POC that works.

**Comment by CyberBotX** [comment]
So personally, I wouldn't be disappointed if TShock dropped arm64 support, but that is only because my use care for using TShock has been on a FreeBSD system, and currently the only version of .NET 6 I've found has been an unofficial build that is x64 only. I have no way of knowing how TShock switching to .NET 6 would affect my running it under FreeBSD, and I've been running it on an x64-based FreeBSD as opposed to an amd64-based FreeBSD mainly because of .NET support under FreeBSD (either through Mono or .NET itself).

**Comment by dioclias** [comment]
I'm hosting on a RaspBerry Pi400 and would be severely disappointed if you stop supporting it, as I won't be able to host a 24/7 server for my friends anymore

**Comment by draGDer** [comment]
Im hosting using Oracle free tier options. The ARM processors are leagues better there. please reconsider dropping support for amr64

**Comment by WinterSolstice8** [comment]
I am currently using a n x84_64 CPU for TShock, but I managed to get an oracle free tier ARM server and it is extremely impressive. I was planning to move my tshock server over there.

In my opinion, if the original Terraria server functions on a platform, I see no reason why TShock shouldn't and I would call it a downgrade if TShock couldn't support the same platforms.

**Comment by GustavoSchip** [comment]
Raspberry pi

**Comment by Mythically** [comment]
[RPi 4B] I am also running a instance of TShock through mono along with a twitch bot that i've written myself and it interacts (wip) with channel point redemptions. I personally thought about running it on my PC, but compared to a RPi it's so inconvenient and wasteful to be honest. I also already had my bot running before starting the server on it, so this is just a great addition to it, 

Unless it's super inefficient,  or it doesn't discourage you from continued development of your fantastic module, i'd like continued support. After all i can just stick to an older version and it should work.  

Thank you for your amazing work so far <3

**Comment by hakusaro** [comment]
I've posted [an update on this situation](https://github.com/Pryaxis/TShock/discussions/2571), but the tl;dr is that based on the feedback here, we're going to keep trying to figure the problem out. Thanks to everyone who responded and provided feedback! It was super helpful!
